File: Simon_Visual_Basic/PruebaLabels 17/PruebaLabels/FrmSimon.vb
Option Strict On

Imports System.Xml
Imports System.Xml.Serialization
Imports System.IO

Public Class FrmSimon


    Private _limite As Integer
    Private _aleatorio As Random
    Private _secuenciaAleatoria As New List(Of Integer)
    Private _indiceSecuencia As Integer
    Private _tiempoRetardo As Integer = 500
    Private _contadorCorrectos As Integer
    Private _numeroUsuario As Integer
    Private _puntaje As Integer
    Public _listaDePuntaje As List(Of Puntaje)

    '   Private _archivoPuntajes As String = "C:\ListaPuntajes.xml"


    Public Sub New()

        ' Llamada necesaria para el Diseñador de Windows Forms.
        InitializeComponent()

        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().
        Me._listaDePuntaje = New List(Of Puntaje)
        Me._contadorCorrectos = 0
        Me._indiceSecuencia = 0
        Me._limite = 0
        Me._puntaje = 0

    End Sub


    Private Sub InicializarLabels()

        Me.LabelEspecial1.BackColor = Color.Blue
        Me.LabelEspecial2.BackColor = Color.DarkGreen
        Me.LabelEspecial3.BackColor = Color.Red
        Me.LabelEspecial4.BackColor = Color.DarkViolet

    End Sub



    Public Sub AgregarManejadoresLabels()

        AddHandler LabelEspecial1.MouseDown, AddressOf LabelsEspecial_MouseDown
        AddHandler LabelEspecial2.MouseDown, AddressOf LabelsEspecial_MouseDown
        AddHandler LabelEspecial3.MouseDown, AddressOf LabelsEspecial_MouseDown
        AddHandler LabelEspecial4.MouseDown, AddressOf LabelsEspecial_MouseDown

    End Sub

    Public Sub QuitarManejadoresLabels()

        RemoveHandler LabelEspecial1.MouseDown, AddressOf LabelsEspecial_MouseDown
        RemoveHandler LabelEspecial2.MouseDown, AddressOf LabelsEspecial_MouseDown
        RemoveHandler LabelEspecial3.MouseDown, AddressOf LabelsEspecial_MouseDown
        RemoveHandler LabelEspecial4.MouseDown, AddressOf LabelsEspecial_MouseDown

    End Sub


    Private Sub CargarSecuenciaAleatoria()

        Me._aleatorio = New Random()

        For i = 0 To 100 'CARGO LA SECUENCIA DE ALEATORIOS CON 100 NUMEROS
            Me._secuenciaAleatoria.Add(Me._aleatorio.Next(0, 4))
        Next

    End Sub


    Private Sub FrmSimon_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load


        Me.LabelEspecial1.ColorEnfocado = Color.Cyan
        Me.LabelEspecial1.ColorDesenfocado = Color.Blue

        Me.LabelEspecial2.ColorEnfocado = Color.LightGreen
        Me.LabelEspecial2.ColorDesenfocado = Color.DarkGreen

        Me.LabelEspecial3.ColorEnfocado = Color.LightPink
        Me.LabelEspecial3.ColorDesenfocado = Color.Red

        Me.LabelEspecial4.ColorEnfocado = Color.Violet
        Me.LabelEspecial4.ColorDesenfocado = Color.DarkViolet


        If File.Exists(_archivoPuntajes) Then
            DeserializarListaPuntajes(_archivoPuntajes)
        End If


        CargarSecuenciaAleatoria()
        CargarListBox()



    End Sub


    Private Sub CargarListBox()

        Me.lstSecuencia.Items.Clear()

        For Each i As Integer In Me._secuenciaAleatoria
            Select Case i
                Case 0
                    Me.lstSecuencia.Items.Add("AZUL")
                Case 1
                    Me.lstSecuencia.Items.Add("VERDE")
                Case 2
                    Me.lstSecuencia.Items.Add("ROJO")
                Case 3
                    Me.lstSecuencia.Items.Add("VIOLETA")
            End Select
        Next

    End Sub



    Private Sub Parpadear(ByVal labelEspecial As Label, ByVal colorEnfocado As Color, ByVal colorDesenfocado As Color)

        labelEspecial.BackColor = colorEnfocado
        labelEspecial.Refresh()
        Threading.Thread.Sleep(Me._tiempoRetardo)

        labelEspecial.BackColor = colorDesenfocado
        labelEspecial.Refresh()
        Threading.Thread.Sleep(Me._tiempoRetardo)

    End Sub




    Private Sub MostrarSecuencia(ByVal sender As System.Object, ByVal e As System.EventArgs)

        QuitarManejadoresLabels()
        InicializarLabels()
        MessageBox.Show("Empieza la secuencia")

        For i = 0 To Me._limite
            Select Case Me._secuenciaAleatoria(i)
                Case 0
                    Parpadear(LabelEspecial1, Color.Cyan, Color.Blue)

                Case 1
                    Parpadear(LabelEspecial2, Color.LightGreen, Color.DarkGreen)

                Case 2
                    Parpadear(LabelEspecial3, Color.LightPink, Color.Red)

                Case 3
                    Parpadear(LabelEspecial4, Color.Violet, Color.DarkViolet)
            End Select

        Next

        MessageBox.Show("Termina la secuencia, TURNO DEL USUARIO")
        AgregarManejadoresLabels()

    End Sub

    Private Sub LabelsEspecial_MouseDown(ByVal sender As System.Object, ByVal e As MouseEventArgs)     '          Handles LabelEspecial1.MouseDown, LabelEspecial2.MouseDown, LabelEspecial3.MouseDown, LabelEspecial4.MouseDown

        Select Case CType(sender, LabelEspecial).Name

            Case "LabelEspecial1"
                Me._numeroUsuario = 0
            Case "LabelEspecial2"
                Me._numeroUsuario = 1
            Case "LabelEspecial3"
                Me._numeroUsuario = 2
            Case "LabelEspecial4"
                Me._numeroUsuario = 3

        End Select

        Comparar(sender, New EventArgs)

    End Sub




    Public Sub Comparar(ByVal sender As Object, ByVal e As EventArgs)

        If _numeroUsuario = Me._secuenciaAleatoria(Me._indiceSecuencia) Then

            '   MessageBox.Show("Correcto")
            Me._contadorCorrectos += 1
            Me._indiceSecuencia += 1

            If Me._contadorCorrectos = Me._limite + 1 Then

                Me._limite += 1
                Me._puntaje += 1 'PUNTAJE VALE LO MISMO QUE LIMITE, SE PUEDE USAR "LIMITE" PARA EL PUNTAJE 
                'EMPIEZA LA SECUENCIA DESDE CERO E INCREMENTA EN UNO EL LIMITE
                Me._indiceSecuencia = 0
                Me._contadorCorrectos = 0

                '  MessageBox.Show("TURNO MAQUINA")
                MostrarSecuencia(sender, e)

            End If
        Else

            MessageBox.Show("Incorrecto")
            GuardarPuntaje()
            '------------PUNTAJE-----------
            Me._listaDePuntaje.Add(New Puntaje(Me._puntaje, Date.Now))
            Me._listaDePuntaje.Add(New Puntaje(Me._limite, Date.Now))
            '-----------------------
            Me._limite = 0
            Me._indiceSecuencia = 0
            Me._contadorCorrectos = 0
            Me._secuenciaAleatoria.RemoveRange(0, Me._secuenciaAleatoria.Count - 1)
            CargarSecuenciaAleatoria()
            CargarListBox()
            MostrarSecuencia(sender, e)

        End If

    End Sub


    Private Sub btnEmpezar_Click(ByVal sender As System.Object, ByVal e As EventArgs) Handles btnEmpezar.Click
        MostrarSecuencia(sender, e)
    End Sub

    Public Sub GuardarPuntaje()
        My.Computer.FileSystem.WriteAllText("C:\PuntosSimon.txt", "Puntos: " & Me._puntaje.ToString() & " - Fecha: " & Date.Now.ToString & vbCrLf, True)
    End Sub


    Private Sub btnMostrarPuntajes_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnMostrarPuntajes.Click

        Dim frm As New FrmPuntajes

        frm.ListaDePuntajes = New List(Of Puntaje)(Me._listaDePuntaje)

        'Text1.Text = FormatNumber(Text1.Text, 2)
        'PARA QUE MUESTRE SOLAMENTE 2 DECIMALES
        frm.lblPromedio.Text = FormatNumber(frm.PromedioGeneral.ToString, 2)
        frm.lblCantPartidos.Text = Me._listaDePuntaje.Count.ToString
        frm.lblMediana.Text = FormatNumber(frm.Mediana.ToString, 2)

        frm.Show()


    End Sub

 

    Public Sub CerrarAplicacion(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles Me.FormClosing

        If MessageBox.Show("Seguro quiere salir?", "Cerrando Aplicacion", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = Windows.Forms.DialogResult.Yes Then

            SerializarListaPuntajes(_archivoPuntajes)

            Me.Dispose()

        Else
            e.Cancel = True

        End If

    End Sub

    Private Function SerializarListaPuntajes(ByVal ruta As String) As Boolean

        Dim pudo As Boolean = False

        Try
            Using escritor As New XmlTextWriter(ruta, System.Text.Encoding.UTF8)

                Dim ser As New XmlSerializer(GetType(List(Of Puntaje)))

                ser.Serialize(escritor, Me._listaDePuntaje)

                pudo = True

            End Using

        Catch ex As Exception

            MessageBox.Show(ex.Message)

        End Try

        Return pudo

    End Function


    Private Function DeserializarListaPuntajes(ByVal ruta As String) As Boolean

        Dim pudo As Boolean = False

        Try
            Using lector As New XmlTextReader(ruta)

                Dim ser As New XmlSerializer(GetType(List(Of Puntaje)))

                Dim listaAux As List(Of Puntaje) = CType(ser.Deserialize(lector), List(Of Puntaje))

                Me._listaDePuntaje = listaAux

                pudo = True

            End Using

        Catch ex As Exception

            MessageBox.Show(ex.Message)

        End Try

        Return pudo

    End Function




End Class

File: Simon_C#/Simon_C_Sharp/FrmSimon.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.IO;
using System.Threading;
using System.Xml.Serialization;
using System.Xml;

namespace Simon_C_Sharp
{
    public partial class FrmSimon : Form 
    {

        private int _limite;
        private Random _aleatorio;
        private List<int> _secuenciaAleatoria;
        private int _indiceSecuencia;
        private int _tiempoRetardo;
        private int _contadorCorrectos;
        private int _numeroUsuario;  
        private int _puntaje;
        private List<Estadisticas> _listaDeEstadisticas;

        public FrmSimon()
        {
            InitializeComponent();

            this._tiempoRetardo = 300;

            this._listaDeEstadisticas = new List<Estadisticas>();
            this._secuenciaAleatoria = new List<int>();

            this.FormClosing += new FormClosingEventHandler(CerrandoAplicacion);

            labelEspecial1.ColorEnfocado = Color.Cyan;
            labelEspecial1.ColorDesenfocado = Color.Blue;

            labelEspecial2.ColorEnfocado = Color.LightGreen;
            labelEspecial2.ColorDesenfocado = Color.DarkGreen;

            labelEspecial3.ColorEnfocado = Color.LightPink;
            labelEspecial3.ColorDesenfocado = Color.Red;

            labelEspecial4.ColorEnfocado = Color.Violet;
            labelEspecial4.ColorDesenfocado = Color.DarkViolet;

            if (File.Exists(Program._archivoPuntajes))
            {
                _listaDeEstadisticas = DeserializarListaEstadisticas(Program._archivoPuntajes);
            }
            CargarSecuenciaAleatoria();
            CargarListBox();

        }

        private void FrmSimon_Load(object sender, EventArgs e)
        {
        }
 
        private void InicializarLabels()
        {
            labelEspecial1.BackColor = Color.Blue;
            labelEspecial2.BackColor = Color.DarkGreen;
            labelEspecial3.BackColor = Color.Red;
            labelEspecial4.BackColor = Color.DarkViolet;
        }

        public void AgregarManejadoresLabels()
        {
            foreach (Control c in this.Controls)
            {
                if (c is LabelEspecial)
                {
                    c.MouseDown += new MouseEventHandler(LabelsEspecial_MouseDown);
                }
            }
        }

        public void QuitarManejadoresLabels()
        {
            foreach (Control c in this.Controls)
            {
                if (c is LabelEspecial)
                {
                    c.MouseDown -= new MouseEventHandler(LabelsEspecial_MouseDown);
                }
            }
        
        }
 
        private void CargarSecuenciaAleatoria()
        {
            this._aleatorio = new Random();

            for (int i = 0; i < 100; i++)
            {
                _secuenciaAleatoria.Add(_aleatorio.Next(0, 4));
            }

        }
 
        private void CargarListBox()
        {
            this.lstSecuencia.Items.Clear();

            foreach (int i in this._secuenciaAleatoria)
            {
                switch (i)
                {
                    case 0:
                        this.lstSecuencia.Items.Add("AZUL");
                        break;
                    case 1:
                        this.lstSecuencia.Items.Add("VERDE");
                        break;
                    case 2:
                        this.lstSecuencia.Items.Add("ROJO");
                        break;
                    case 3:
                        this.lstSecuencia.Items.Add("VIOLETA");
                        break;
                }
            }  
        }  


        private void Parpadear(LabelEspecial labelEspecial, Color colorEnfocado, Color colorDesenfocado)
        {
            labelEspecial.BackColor = colorEnfocado;
            labelEspecial.Refresh();
            Thread.Sleep(this._tiempoRetardo);

            labelEspecial.BackColor = colorDesenfocado;
            labelEspecial.Refresh();
            Thread.Sleep(this._tiempoRetardo);
        }
 
        public void MostrarSecuencia(Object sender, EventArgs e)
        {
            QuitarManejadoresLabels();
            InicializarLabels();
            MessageBox.Show("Empieza la secuencia");

            for (int i = 0; i <= _limite; i++)
            {
                if (_secuenciaAleatoria[i] == 0) Parpadear(labelEspecial1, this.labelEspecial1.ColorEnfocado, this.labelEspecial1.ColorDesenfocado);
                if (_secuenciaAleatoria[i] == 1) Parpadear(labelEspecial2, this.labelEspecial2.ColorEnfocado, this.labelEspecial2.ColorDesenfocado);
                if (_secuenciaAleatoria[i] == 2) Parpadear(labelEspecial3, this.labelEspecial3.ColorEnfocado, this.labelEspecial3.ColorDesenfocado);
                if (_secuenciaAleatoria[i] == 3) Parpadear(labelEspecial4, this.labelEspecial4.ColorEnfocado, this.labelEspecial4.ColorDesenfocado);
            }
            MessageBox.Show("Termina la secuencia, TURNO DEL USUARIO");
            AgregarManejadoresLabels();
        }

             
        private void LabelsEspecial_MouseDown(Object sender, MouseEventArgs e)
        {
            switch (((LabelEspecial)sender).Name)
            {
                case "labelEspecial1":
                    _numeroUsuario = 0; break;
                case "labelEspecial2":
                    _numeroUsuario = 1; break;
                case "labelEspecial3":
                    _numeroUsuario = 2; break;
                case "labelEspecial4":
                    _numeroUsuario = 3; break;
            }
 
            Comparar(sender, new EventArgs());   
        }//----------------

        public void Comparar(Object sender, EventArgs e)
        {
            if (_numeroUsuario == this._secuenciaAleatoria[_indiceSecuencia])
            {
                _contadorCorrectos++;
                _indiceSecuencia++;

                if (_contadorCorrectos == _limite + 1)
                {
                    _limite++;
                    _puntaje++;

                    _indiceSecuencia = 0;
                    _contadorCorrectos = 0;
                    // MessageBox.Show("TURNO MAQUINA")
                    MostrarSecuencia(sender, e);
                }//fin if
            }   
            else
            {
                MessageBox.Show("Incorrecto");
                GuardarPuntaje();

                //------------PUNTAJE-----------
              // this._listaDeEstadisticas .Add(new Estadisticas(_puntaje, DateTime.Now));
               this._listaDeEstadisticas.Add(new Estadisticas(_limite, DateTime.Now));
                //-----------------------

                _limite = 0;
                _indiceSecuencia = 0;
                _contadorCorrectos = 0;
                _secuenciaAleatoria.RemoveRange(0, _secuenciaAleatoria.Count - 1);
                CargarSecuenciaAleatoria();
                CargarListBox();
                MostrarSecuencia(sender, e);

            }
        } //---    


        private void btnEmpezar_Click_1(Object sender, EventArgs e)
        {
            MostrarSecuencia(sender, e);
        }
         
 
        private void GuardarPuntaje()
        {
            // My.Computer.FileSystem.WriteAllText("C:\PuntosSimon.txt", "Puntos: " & Me._puntaje.ToString() & " - Fecha: " & Date.Now.ToString & vbCrLf, True)
        }
   
        private bool SerializarListaPuntajes(String ruta)
        {
            bool pudo = false;
            try
            {
                using (XmlTextWriter escritor =
                    new XmlTextWriter(ruta, Encoding.UTF8))
                {
                    XmlSerializer ser = new XmlSerializer(typeof(List<Estadisticas>));
                    ser.Serialize(escritor, this._listaDeEstadisticas);
                    pudo = true;
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
            return pudo;
        }

        public static List<Estadisticas> DeserializarListaEstadisticas(String ruta)
        {
            List<Estadisticas> listaAux = new List<Estadisticas>();
            try
            {
                using (XmlTextReader lector = new XmlTextReader(ruta))
                {
                    XmlSerializer ser =
                        new XmlSerializer(typeof(List<Estadisticas>));
                    listaAux = (List<Estadisticas>)ser.Deserialize(lector);
                    //this._listaDeEstadisticas = listaAux;
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
            return listaAux;
        }

        public void CerrandoAplicacion(Object sender, FormClosingEventArgs e)
        {
            if (MessageBox.Show("Seguro quiere salir ?", "Cerrando aplicacion",
                MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes)
            {
                 SerializarListaPuntajes(Program._archivoPuntajes);
              //  this.Dispose();
            }
            else
            {
                e.Cancel = true;
            }
        }

 
    } //----------------------------------------------------------------
}
 